x-pack/metricbeat/module/azure/billing/data.go

```diff
@@ -114,44 +121,134 @@ func EventsMapping(subscriptionId string, results Usage, startTime time.Time, en
 	//
 	// Forecasts
 	//
-	groupedCosts := make(map[*string][]consumption.Forecast)
 
-	for _, forecast := range results.ForecastCosts {
-		groupedCosts[forecast.UsageDate] = append(groupedCosts[forecast.UsageDate], forecast)
+	forecastsEvents, err := getEventsFromQueryResult(results.Forecasts, subscriptionId)
+	if err != nil {
+		return events, err
 	}
 
-	for _, forecast := range results.ActualCosts {
-		groupedCosts[forecast.UsageDate] = append(groupedCosts[forecast.UsageDate], forecast)
+	events = append(events, forecastsEvents...)
+
+	return events, nil
+}
+
+// getResourceNameFromPath returns the resource name by picking the last part from a `/` separated resource path.
+//
+// For example, given a path like the following:
+// `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachines/{vmName}`
+//
+// It would return the value `{vmName}`.
+func getResourceNameFromPath(path string) string {
+	parts := strings.Split(path, "/")
+	// According to the documentation, `string.Split()` always returns a non-empty slice when the separator is not empty,
+	// so it should be safe to use `len(parts) - 1` to get the last element.
+	return parts[len(parts)-1]
+}
+
+// getEventsFromQueryResult returns the events from the QueryResult obtained Cost Management API.
+//
+// Here's what you'll find in the QueryResult:
+//
+// .Columns:
+// 0: Cost: Number
+// 1: UsageDate: Number
+// 2: CostStatus: String
+// 3: Currency: String
+//
+// .Rows:
+// 0: []interface {}{0.11, 2.0200807e+07, "Actual", "USD"}
+// 1: []interface {}{0.11, 2.0200808e+07, "Forecast", "USD"}
+//
+func getEventsFromQueryResult(result costmanagement.QueryResult, subscriptionID string) ([]mb.Event, error) {
+	// The number of columns expected in the QueryResult supported by this input.
+	// The structure of the QueryResult is determined by the value we set in
+	// the `costmanagement.ForecastDefinition` struct at query time.
+	const expectedNumberOfColumns = 4
+
+	if result.Columns == nil || len(*result.Columns) != expectedNumberOfColumns {
+		return []mb.Event{}, fmt.Errorf("unsupported forecasts QueryResult format: %d instead of %d", len(*result.Columns), expectedNumberOfColumns)
 	}
 
-	for usageDate, items := range groupedCosts {
-		var actualCost *decimal.Decimal
-		var forecastCost *decimal.Decimal
-		for _, item := range items {
-			if item.ChargeType == consumption.ChargeTypeActual {
-				actualCost = item.Charge
-			} else {
-				forecastCost = item.Charge
+	if result.Rows == nil {
+		return []mb.Event{}, errors.New("unsupported forecasts QueryResult format: no rows")
+	}
+
+	events := make([]mb.Event, 0, len(*result.Rows))
+	for _, row := range *result.Rows {
+		var cost float64
+		var currency string
+		var costStatus string
+		var usageDate time.Time
+
+		if len(row) != expectedNumberOfColumns {
+			return events, fmt.Errorf("unsupported forecasts QueryResult.Rows format: %d instead of %d", len(row), expectedNumberOfColumns)
```

Comment:
do we want to return an error and skip collecting forecast events or just skip the rows that does not pass validation?

Reply:
Yep, good question indeed.

During the development, we never saw rouge rows that were not conforming with the expected format. The resulting structure is dictated by the [costmanagement.ForecastDefinition](https://github.com/elastic/beats/blob/45b2a55d3576edc60be1c443c615c6a5e0aa4fa1/x-pack/metricbeat/module/azure/billing/service.go#L90-L114) struct, in particular, `Aggregation` and `Granularity`.

Checking the number of columns is a simple safeguard mechanism to avoid crashes when we change the `costmanagement.ForecastDefinition` or upgrade the Cost Management API.

I think skipping the record with a proper log is a better approach to returning the error: I'll change it!

Reply:
Updated in https://github.com/elastic/beats/pull/32589/commits/a23610072d3caf29e98d57d68695380b3e29b5ab

Reply:
👍